### ISSUE TITLE
Prevent issuing an error on type placeholders not present in user code

### ISSUE THREAD

**Issue by cramertj** [rust-lang/rust#54536]
The following function issues two separate errors-- one for the user's mistake ("parenthesized parameters may only be used with a trait") and one bogus error ("the type placeholder `_` is not allowed within types on item signatures").

```rust
fn foo() -> Result((&str, i32), i32) {}
```

Errors:

```
error[E0214]: parenthesized parameters may only be used with a trait
 --> src/main.rs:1:19
  |
1 | fn foo() -> Result((&str, i32), i32) {}
  |                   ^^^^^^^^^^^^^^^^^^ only traits may use parentheses

error[E0121]: the type placeholder `_` is not allowed within types on item signatures
 --> src/main.rs:1:13
  |
1 | fn foo() -> Result((&str, i32), i32) {}
  |             ^^^^^^^^^^^^^^^^^^^^^^^^ not allowed in type signatures

error: aborting due to 2 previous errors
```

The second error is misleading and unhelpful here since the user didn't write a type placeholder.

**Comment by estebank** [comment]
Current output is much better:

```
error[E0214]: parenthesized type parameters may only be used with a `Fn` trait
 --> src/lib.rs:1:19
  |
1 | fn foo() -> Result((&str, i32), i32) {}
  |                   ^^^^^^^^^^^^^^^^^^
  |                   |
  |                   only `Fn` traits may use parentheses
  |                   help: use angle brackets instead: `<(&str, i32), i32>`

error[E0106]: missing lifetime specifier
 --> src/lib.rs:1:21
  |
1 | fn foo() -> Result((&str, i32), i32) {}
  |                     ^ help: consider giving it a 'static lifetime: `&'static`
  |
  = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
```
